apps/wing/src/commands/test/snapshots.ts

```diff
@@ -0,0 +1,178 @@
+import * as cp from "child_process";
+import { existsSync, readFileSync, writeFileSync } from "fs";
+import * as fs from "fs/promises";
+import { join, extname, basename } from "path";
+import { BuiltinPlatform } from "@winglang/compiler";
+import * as glob from "glob";
+import { SNAPSHOTS_ERROR_HELP } from "./snapshots-help";
+import { TestOptions } from "./test";
+import { renderTestName } from "./util";
+import { withSpinner } from "../../util";
+import { compile } from "../compile";
+
+export enum SnapshotMode {
+  /**
+   * Auto-detect. If CI=1, SnapshotMode.ASSERT otherwise SnapshotMode.DEPLOY.
+   */
+  AUTO = "auto",
+
+  /**
+   * Snapshots are disabled.
+   */
+  NEVER = "never",
+
+  /**
+   * Deploy and test in the cloud and update the snapshot only if all tests passed.
+   */
+  DEPLOY = "deploy",
+
+  /**
+   * Update the snapshots without deploying (dry run).
+   */
+  UPDATE = "update",
+
+  /**
+   * Just verify that the snapshots are correct.
+   */
+  ASSERT = "assert",
+}
+
+export function determineSnapshotMode(
+  target: string | undefined,
+  options: TestOptions
+): SnapshotMode {
+  // no snapshots for sim, ever!
+  if (target === BuiltinPlatform.SIM) {
+    return SnapshotMode.NEVER;
+  }
+
+  // if AUTO or unset, determine behavior based on CI flag
+  if (!options.snapshots || options.snapshots === SnapshotMode.AUTO) {
+    // determine behavior based on CI
+    if (process.env.CI) {
+      return SnapshotMode.ASSERT;
+    } else {
+      return SnapshotMode.DEPLOY;
+    }
+  }
+
+  return options.snapshots;
+}
+
+export async function captureSnapshot(entrypoint: string, target: string, options: TestOptions) {
+  const snapshotMode = determineSnapshotMode(target, options);
+
+  // skip if snapshots are disabled
+  if (snapshotMode === SnapshotMode.NEVER) {
+    return;
+  }
+
+  const snapshotFile = `${entrypoint}.${target}.snap.md`;
+
+  await withSpinner(`Snapshot ${renderTestName(snapshotFile)}...`, async () => {
+    // we need to compile again in order for snapshots because we can't afford the snapshot to be
+    // based on a random root id (which is the default for tests)
+    const snapshotDir = await compile(entrypoint, options);
```

Comment:
Should the `testing` compiler option also be set to true like it is for the test compilation.

Reply:
No, because this will result in non deterministic output (test apps have a random root for isolation).

Reply:
[Here](https://github.com/winglang/wing/blob/main/apps/wing/src/commands/test/test.ts#L149-L153) both the random root id is set and a testing prop is true.
I understand avoiding the random root id for this, but the other prop seems important as well.

Reply:
I am not sure I understand. Why does it make sense to set `testing` to `true` in this context? We are not really running the test, just synthesizing the output. I rather the output be as true to the production output as possible.

Reply:
> We are not really running the test, just synthesizing the output.

Setting testing to true sets the WING_IS_TEST env to true, which sets isTestEnvironment in the app to true.

This flag is used in several places before running the tests, including being exposed in actual wing code. So it changes the resulting compilation.

> I rather the output be as true to the production output as possible.

The file being compiled could be a .test.w file, which means it probably doesn't have "production" output. `test {}` blocks are not included in production output. I figured the snapshot is reflecting the tests infrastructure rather than what'd you'd get from running "wing compile -t tf-aws entrypoint". The production infra isn't useful here because that's not what the test is using.